### DISCUSSION TITLE
tutorials/intro-typescript/10-querying-real-data

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#149] [course feedback]
# tutorials/intro-typescript/10-querying-real-data

https://www.apollographql.com/tutorials/intro-typescript/10-querying-real-data

**Comment by atharrison** [comment]
When I got to the step to inspect the dataSources in resolvers.ts, I got the following error:

  ```
  > Intro to GraphQL with TypeScript@0.0.0 dev
> ts-node-dev --respawn ./src/index.ts

[INFO] 18:18:00 ts-node-dev ver. 2.0.0 (using ts-node ver. 10.9.2, typescript ver. 5.3.3)
Compilation error in /Users/andrew/workspace/apollo/odyssey-intro-typescript/src/resolvers.ts
[ERROR] 18:18:01 ⨯ Unable to compile TypeScript:
src/resolvers.ts(6,9): error TS2322: Type '(_: {}, __: {}, { dataSources }: DataSourceContext) => void' is not assignable to type 'Resolver<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, {}>'.
  Type '(_: {}, __: {}, { dataSources }: DataSourceContext) => void' is not assignable to type 'ResolverFn<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, {}>'.
    Type 'void' is not assignable to type 'ResolverTypeWrapper<Listing>[] | Promise<ResolverTypeWrapper<Listing>[]>'.
    ```

**Reply by atharrison** [reply]
Ah, I found the issue. There is a missing step where the `getFeaturedListings` method in `src/datasources/listing-api.ts` needs to be changed from `void` to `Promise<Listing[]>`.

The function now looks like:

```
getFeaturedListings(): Promise<Listing[]> {
        return this.get<Listing[]>("featured-listings");
    }
```

**Reply by lizhennessy** [reply]
Ah thanks for calling this out @atharrison ! I'll make this step in [Lesson 9](https://www.apollographql.com/tutorials/intro-typescript/09-codegen) a little bit easier to catch! (It _is_ a bit buried under all the codegen content!) Let us know if you run into any other problems. :)

**Comment by bijayabc** [comment]
I am getting the following error because of some error in the resolvers.ts file:
Type '(_: {}, __: {}, { dataSources }: DataSourceContext) => () => Promise<Listing[]>' is not assignable to type 'Resolver<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, {}>'.
  Type '(_: {}, __: {}, { dataSources }: DataSourceContext) => () => Promise<Listing[]>' is not assignable to type 'ResolverFn<ResolverTypeWrapper<Listing>[], {}, DataSourceContext, {}>'.
    Type '() => Promise<Listing[]>' is not assignable to type 'ResolverTypeWrapper<Listing>[] | Promise<ResolverTypeWrapper<Listing>[]>'.ts(2322)
resolvers.ts(5, 27): Did you mean to call this expression?
(parameter) dataSources: {
    listingAPI: ListingAPI;
}

**Reply by lizhennessy** [reply]
Hi there @bijayabc ! i'm happy to take a look. Do you have your code pushed up to a repository I can access?

**Reply by subhash-malireddy** [reply]
@bijayabc ~it seems like you are directly providing the class itself without instantiating it. 
Update the dataSources in the following way~

~dataSources: {
    listingAPI: new ListingAPI({ cache }),
}~


Sorry for my oversight.. it seems like you are returning a function rather than returning a promise of type Listings[] within your featuredListings resolver function. 

Check your return statement it should be like this, `return dataSources.listingAPI.getListings()`

**Comment by joaquinm-omf** [comment]
I'm getting this response when trying to query the data, any idea?


{
  "errors": [
    {
      "message": "request to https://rt-airlock-services-listing.herokuapp.com/featured-listings failed, reason: unable to get local issuer certificate",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "featuredListings"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "FetchError: request to https://rt-airlock-services-listing.herokuapp.com/featured-listings failed, reason: unable to get local issuer certificate",
          "    at ClientRequest.<anonymous> (/Users/joaquin.martinezvaldepenas/node_modules/node-fetch/lib/index.js:1501:11)",
          "    at ClientRequest.emit (node:events:508:28)",
          "    at emitErrorEvent (node:_http_client:108:11)",
          "    at TLSSocket.socketErrorListener (node:_http_client:575:5)",
          "    at TLSSocket.emit (node:events:508:28)",
          "    at emitErrorNT (node:internal/streams/destroy:170:8)",
          "    at emitErrorCloseNT (node:internal/streams/destroy:129:3)",
          "    at processTicksAndRejections (node:internal/process/task_queues:89:21)"
        ]
      }
    }
  ],
